File: DCC2024_Artur/January/Day3_Number_of_Laser_Beams_in_a_Bank.cpp
/*
 https://leetcode.com/problems/number-of-laser-beams-in-a-bank/?envType=daily-question&envId=2024-01-03
*/

#include <iostream>
#include "vector"
#include "string"
using namespace std;

class Solution
{
public:
    int numberOfBeams(vector<string>& bank)
    {
        int prevLasers = 0;
        int beam = 0;

        for(const auto& row : bank)
        {
            int currLasers = 0;
            for(auto cell : row)
                if(cell == '1')
                    currLasers++;
            if(currLasers)
            {
                beam += prevLasers * currLasers;
                prevLasers = currLasers;
            }
        }

        return beam;
    }
};

int main()
{
    vector<string> test = {"011001","000000","010100","001000"};
    Solution solution;

    cout << solution.numberOfBeams(test);
}

/*
 Main idea:
 Є три змінні prevLasers — кількість лазерів в останньому ряді, де
 були лазери, currLasers — кількість лазерів у поточному ряді та
 beam — загальна кількість променів у банку.

 Проходитись по кожному рядку, рахуючи кількість одиниць(лазерів).
 Якщо їхня кількість відмінна від 0, то до beam додаємо добуток
 кількості лазерів на поточному та в останньому ряді у якому були
 лазери. Перезаписуємо значення prevLasers на currLasers.

 Чому prevLasers не обов'язково попередній ряд?
 Усе тому, що за умовою лазери можуть утворювати промінь через
 декілька рядів, за умовою, що на них не було жодного іншого лазера

 Приклад:

 011100
 000000
 000000
 010010
 000100

 Тут лазери з першого ряду будуть утворювати промені із лазерами з 4
 ряду, бо в 2 та 3 рядах немає жодного лазера. Проте 1 ряд не може
 утворювати промені з 5 рядом, оскільки між ними в 4 ряді розташовані
 2 лазера.
 */

File: DCC2024_Artur/January/Day3_Number_of_Laser_Beams_in_a_Bank.py
"""
https://leetcode.com/problems/number-of-laser-beams-in-a-bank/?envType=daily-question&envId=2024-01-03
"""

from typing import List

class Solution(object):
    def numberOfBeams(self, bank: List[str]) -> int:
        """
        :type bank: List[str]
        :rtype: int
        """
        prevRow = 0
        beam = 0
        
        for row in bank:
            currRow = row.count('1')
            beam += prevRow * currRow
            prevRow = currRow if currRow else prevRow
            
        return beam


if (__name__ == '__main__'):
    solution = Solution()
    test = ["011001","000000","010100","001000"]
    print(solution.numberOfBeams(test))


"""
 Ідея описана в С++ файлі.
 В імплементації є відмінності, замість ще одного циклу
 із підрахункому одиниць в рядку використовується вбудований
 метод count()/.
"""
